server/src/main/java/com/linecorp/centraldogma/server/internal/storage/RepositoryGarbageCollectingServicePlugin.java

```diff
@@ -0,0 +1,225 @@
+        scheduledFuture = gcWorker.schedule(() -> gc(context), nextExecution);
+
+        Futures.addCallback(scheduledFuture, new FutureCallback<Object>() {
+            @Override
+            public void onSuccess(@Nullable Object result) {}
+
+            @Override
+            public void onFailure(Throwable cause) {
+                if (!stopping) {
+                    logger.warn("Repository gc scheduler stopped due to an unexpected exception:", cause);
+                }
+            }
+        }, gcWorker);
+    }
+
+    @Override
+    public synchronized CompletionStage<Void> stop(PluginContext context) {
+        stopping = true;
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(false);
+        }
+
+        try {
+            if (gcWorker != null && !gcWorker.isTerminated()) {
+                logger.info("Stopping the repository gc worker ..");
+                boolean interruptLater = false;
+                while (!gcWorker.isTerminated()) {
+                    gcWorker.shutdownNow();
+                    try {
+                        gcWorker.awaitTermination(1, TimeUnit.SECONDS);
+                    } catch (InterruptedException e) {
+                        // Interrupt later.
+                        interruptLater = true;
+                    }
+                }
+                logger.info("Stopped the repository gc worker.");
+
+                if (interruptLater) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        } catch (Throwable t) {
+            logger.warn("Failed to stop the repository gc worker:", t);
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @VisibleForTesting
+    void gc(PluginContext context) {
+        if (stopping) {
+            return;
+        }
+
+        final ProjectManager pm = context.projectManager();
+        final Stopwatch stopwatch = Stopwatch.createUnstarted();
+        for (Project project : pm.list().values()) {
+            for (Repository repo : project.repos().list().values()) {
+                runGc(project, repo, stopwatch);
+            }
+
+            runGc(project, project.metaRepo(), stopwatch);
+
+            // Clean up the revision caches of repositories removed.
+            for (final String removed : project.repos().listRemoved().keySet()) {
+                gcRevisions.remove(project.name() + '/' + removed);
+            }
+        }
+
+        scheduleGc(context);
+    }
+
+    private void runGc(Project project, Repository repo, Stopwatch stopwatch) {
+        try {
+            final String cacheKey = project.name() + '/' + repo.name();
+            if (!needsGc(repo, cacheKey)) {
+                return;
+            }
+
+            logger.info("Starting repository gc on {}/{} ..", project.name(), repo.name());
+            stopwatch.reset();
+            final Revision gcRevision = repo.gc();
+            gcRevisions.put(cacheKey, gcRevision);
```

Comment:
Shouldn't we store this into the hard disk? (e.g. metadata.json)

Reply:
I thought about that. But it is weird that a gc creates an additional git commit. Would it better to write to a separate file?

Reply:
Added `GitGcRevision` for that purpose. [`db5dc16` (#564)](https://github.com/line/centraldogma/pull/564/commits/db5dc16c99fcfc99ad7b24131478feae5efa9b1a)